x-pack/plugins/security/server/authentication/api_keys.ts

```diff
@@ -131,6 +155,39 @@ export class APIKeys {
     return result;
   }
 
+  /**
+   * Tries to grant an API key for the current user.
+   * @param request Request instance.
+   */
+  async grantAsInternalUser(request: KibanaRequest): Promise<GrantAPIKeyResult | null> {
+    if (!this.license.isEnabled()) {
+      return null;
+    }
+
+    this.logger.debug('Trying to grant an API key');
+    const authorizationHeader = HTTPAuthorizationHeader.parseFromRequest(request);
```

Comment:
What you've said makes sense. I'm also open to alternatives which require some changes to the authentication providers. I began exploring what it'd look like to allow the auth providers to store the parameters we need to grant an API keyu in a `WeakMap<KibanaRequest, GrantAPIKeyParams>` that only the security plugin had access to, but then came on the necessity to parse the `Authorization` header within the `http` authentication provider and decided to just do the header parsing logic here.

Reply:
Yeah, I think what you have now is the best option for the time being. Parsing headers because of `http` is where I ended up as well when we had initial discussion on that requirement.

It feels we're still missing some pieces either in Security plugin or in the Core that would have allowed us to come with a more future-proof approach, but I'm sure we'll get there and reuse the work you've done here anyway.